README.rst
===================
awsenergylabelerlib
===================

Project energy labeling accounts and landing zone based on findings of Security Hub in AWS cloud.


* Documentation: https://awsenergylabelerlib.readthedocs.org/en/latest


Development Workflow
====================

The workflow supports the following steps

 * lint
 * test
 * build
 * document
 * upload
 * graph

These actions are supported out of the box by the corresponding scripts under _CI/scripts directory with sane defaults based on best practices.
Sourcing setup_aliases.ps1 for windows powershell or setup_aliases.sh in bash on Mac or Linux will provide with handy aliases for the shell of all those commands prepended with an underscore.

The bootstrap script creates a .venv directory inside the project directory hosting the virtual environment. It uses pipenv for that.
It is called by all other scripts before they do anything. So one could simple start by calling _lint and that would set up everything before it tried to actually lint the project

Once the code is ready to be delivered the _tag script should be called accepting one of three arguments, patch, minor, major following the semantic versioning scheme.
So for the initial delivery one would call

    $ _tag --minor

which would bump the version of the project to 0.1.0 tag it in git and do a push and also ask for the change and automagically update HISTORY.rst with the version and the change provided.


So the full workflow after git is initialized is:

 * repeat as necessary (of course it could be test - code - lint :) )

   * code
   * lint
   * test
 * commit and push
 * develop more through the code-lint-test cycle
 * tag (with the appropriate argument)
 * build
 * upload (if you want to host your package in pypi)
 * document (of course this could be run at any point)


Important Information
=====================

This template is based on pipenv. In order to be compatible with requirements.txt so the actual created package can be used by any part of the existing python ecosystem some hacks were needed.
So when building a package out of this **do not** simple call

    $ python setup.py sdist bdist_egg

**as this will produce an unusable artifact with files missing.**
Instead use the provided build and upload scripts that create all the necessary files in the artifact.



Project Features
================

* TODO


HISTORY.rst
.. :changelog:

History
-------

0.0.1 (09-11-2021)
---------------------

* First code creation


0.1.0 (09-11-2021)
------------------

* Initial pypi release.


0.1.1 (09-11-2021)
------------------

* Exposed main object to the root of the package.


0.2.0 (26-11-2021)
------------------

* Fixed labaling algorithms, added retries to finding retrieval and implemented proper exception handling.


0.2.1 (02-12-2021)
------------------

* Updated number of days open to 999 days
* Account data now includes details on number of findings per severity


0.2.2 (06-12-2021)
------------------

* Improved error handling
* Allow/deny specific regions


0.2.3 (06-12-2021)
------------------

* Added requests dependency


0.2.4 (10-12-2021)
------------------

* A finding exposes now more fields: resources, record_state, description, remediation


0.2.5 (10-12-2021)
------------------

* Compliance fields added to a Security Hub Finding


0.3.0 (14-12-2021)
------------------

* Changed the structure of the measurement data


0.4.0 (28-01-2022)
------------------

* Introduced single account mode which opportunistically gets findings from SecurityHub


0.4.1 (01-02-2022)
------------------

* Edited the filter to only include FAILED findings so NOT_AVAILABLE aren't counted as findings anymore


0.4.2 (02-03-2022)
------------------

* Suppressed findings are no longer counted into the calculation.
* Framework validation works as expected now.


0.4.3 (04-03-2022)
------------------

* Filtered out Archived findings.


0.4.4 (04-03-2022)
------------------

* Filtered out archived findings.


0.4.5 (04-03-2022)
------------------

* No duplicates anymore, unique findings only.


1.0.0 (14-03-2022)
------------------

* Complete redesign of the api and many optimisations of retrieving findings and calculating labels.


1.0.1 (14-03-2022)
------------------

* Cached some calculations to prevent duplications and standardized on default labels with properly retrieved values.


1.1.0 (17-03-2022)
------------------

* Implemented client side finding filtering.


1.1.1 (19-05-2022)
------------------

* Fix to strip leading slash in S3 destination path


1.1.2 (23-08-2022)
------------------

* Fix bug related to exporting resources data


1.2.0 (09-09-2022)
------------------

* Removed pandas dependency in favor of native python functionality.


1.2.1 (26-09-2022)
------------------

* Fixed timestamp bug
* Fixed bug where accounts without findings got an F


1.2.2 (26-09-2022)
------------------

* Fixed timestamp bug
* Fixed bug where accounts without findings got an F


2.0.0 (25-10-2022)
------------------

* Removed "CIS" from default frameworks, fixed a bug with required region for security hub service.


2.0.1 (25-10-2022)
------------------

* Set appropriate logging level for unsuccessful retrieval of account alias.


2.0.2 (02-11-2022)
------------------

* Fixes parsing for dates for inspector findings by implementing auto datetime parsing.


3.0.0 (02-11-2022)
------------------

* Implements the concept of a Zone to not require full Organizations access rights.


3.1.0 (16-11-2022)
------------------

* Implemented support for metadata, seperated critical and high findings in the report and implemented support for CIS 1.4 findings.


3.2.0 (06-03-2023)
------------------

* Bump dependencies.


3.2.1 (10-03-2023)
------------------

* Updated dependencies


3.2.2 (11-04-2023)
------------------

* Convert Dataclasses to Frozen for compatibility with latest python.


3.2.3 (13-04-2023)
------------------

* Changed from dataclass to normal.


3.2.4 (24-04-2023)
------------------

* Writing to S3 now does not attempt to write to a temporary file first, accomodating read-only filesystems when only writing to S3.


4.0.0 (06-06-2023)
------------------

* Implement support for finding consolidation.


5.0.0 (01-12-2023)
------------------

* Remove alias functionality since it was not working properly.


5.0.1 (01-12-2023)
------------------

* Fix upload mechanism and bump dependency versions.


awsenergylabelerlib/configuration.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: configuration.py
#
# Copyright 2021 Costas Tyfoxylos, Jenda Brands, Theodoor Scholte
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
configuration package.

Import all parts from configuration here

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html
"""

import json
import logging
import urllib.error
import urllib.request

from .awsenergylabelerlibexceptions import UnableToRetrieveSecurityHubRegions
from .datamodels import (ZoneEnergyLabelingData,
                         SecurityHubFindingsData,
                         SecurityHubFindingsResourcesData,
                         SecurityHubFindingsTypesData,
                         LabeledAccountsData,
                         LabeledAccountData,
                         Metadata)

__author__ = 'Costas Tyfoxylos <ctyfoxylos@schubergphilis.com>'
__docformat__ = '''google'''
__date__ = '''09-11-2021'''
__copyright__ = '''Copyright 2021, Costas Tyfoxylos, Jenda Brands, Theodoor Scholte'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<ctyfoxylos@schubergphilis.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".

LOGGER_BASENAME = '''configuration'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())

ACCOUNT_THRESHOLDS = [{'label': 'A',
                       'critical_high': 0,
                       'medium': 10,
                       'low': 20,
                       'days_open_less_than': 999},
                      {'label': 'B',
                       'critical_high': 10,
                       'medium': 20,
                       'low': 40,
                       'days_open_less_than': 999},
                      {'label': 'C',
                       'critical_high': 15,
                       'medium': 30,
                       'low': 60,
                       'days_open_less_than': 999},
                      {'label': 'D',
                       'critical_high': 20,
                       'medium': 40,
                       'low': 80,
                       'days_open_less_than': 999},
                      {'label': 'E',
                       'critical_high': 25,
                       'medium': 50,
                       'low': 100,
                       'days_open_less_than': 999}]

ZONE_THRESHOLDS = [{'label': 'A',
                    'percentage': 90},
                   {'label': 'B',
                    'percentage': 70},
                   {'label': 'C',
                    'percentage': 50},
                   {'label': 'D',
                    'percentage': 30},
                   {'label': 'E',
                    'percentage': 20}]

DEFAULT_SECURITY_HUB_FILTER = {'UpdatedAt': [{'DateRange': {'Value': 7,
                                                            'Unit': 'DAYS'}}],
                               'ComplianceStatus': [{'Value': 'FAILED',
                                                     'Comparison': 'EQUALS'}],
                               'WorkflowStatus': [{'Value': 'SUPPRESSED',
                                                   'Comparison': 'NOT_EQUALS'}],
                               'RecordState': [{'Value': 'ARCHIVED',
                                                'Comparison': 'NOT_EQUALS'}]}

AWS_FOUNDATIONAL_SECURITY_FRAMEWORK = 'aws-foundational-security-best-practices'
CIS_AWS_FOUNDATION_FRAMEWORK = 'cis-aws-foundations-benchmark'
NIST_800_53_FRAMEWORK = 'nist-800-53'
PCI_DSS_FRAMEWORK = 'pci-dss'
DEFAULT_SECURITY_HUB_FRAMEWORKS = {AWS_FOUNDATIONAL_SECURITY_FRAMEWORK}


def get_available_security_hub_regions():
    """The regions that security hub can be active in.

    Returns:
        regions (list): A list of strings of the regions that security hub can be active in.

    """
    url = 'https://api.regional-table.region-services.aws.a2z.com/index.json'
    try:
        with urllib.request.urlopen(url) as response:
            response_json = json.loads(response.read())
    except (urllib.error.URLError, ValueError):
        raise UnableToRetrieveSecurityHubRegions('Failed to retrieve applicable AWS regions') from None
    return [entry.get('id', '').split(':')[1]
            for entry in response_json.get('prices')
            if entry.get('id').startswith('securityhub')]


SECURITY_HUB_ACTIVE_REGIONS = ['ap-east-1', 'ap-northeast-2', 'ap-southeast-1', 'ap-southeast-2', 'ca-central-1',
                               'eu-north-1', 'eu-west-2', 'us-east-2', 'us-gov-west-1', 'us-west-2', 'af-south-1',
                               'ap-northeast-3', 'cn-northwest-1', 'eu-south-1', 'eu-west-1', 'eu-west-3', 'me-south-1',
                               'sa-east-1', 'us-east-1', 'us-west-1', 'ap-northeast-1', 'ap-south-1', 'cn-north-1',
                               'eu-central-1', 'us-gov-east-1']

FILE_EXPORT_TYPES = [
    {'type': 'zone_energy_label',
     'filename': 'zone-energy-label.json',
     'object_type': ZoneEnergyLabelingData,
     'required_arguments': ['name', 'energy_label']},
    {'type': 'findings',
     'filename': 'security-hub-findings.json',
     'object_type': SecurityHubFindingsData,
     'required_arguments': ['security_hub_findings']},
    {'type': 'findings_resources',
     'filename': 'security-hub-findings-resources.json',
     'object_type': SecurityHubFindingsResourcesData,
     'required_arguments': ['security_hub_findings']},
    {'type': 'findings_types',
     'filename': 'security-hub-findings-types.json',
     'object_type': SecurityHubFindingsTypesData,
     'required_arguments': ['security_hub_findings']},
    {'type': 'labeled_accounts',
     'filename': 'labeled-accounts.json',
     'object_type': LabeledAccountsData,
     'required_arguments': ['labeled_accounts']},
    {'type': 'account_energy_label',
     'filename': 'account-energy-label.json',
     'object_type': LabeledAccountData,
     'required_arguments': ['labeled_accounts']},
    {'type': 'metadata',
     'filename': 'metadata.json',
     'object_type': Metadata,
     'required_arguments': ['metadata']},
]

ZONE_METRIC_EXPORT_TYPES = ['zone_energy_label', 'labeled_accounts', 'metadata']

ACCOUNT_METRIC_EXPORT_TYPES = ['account_energy_label', 'metadata']

DATA_EXPORT_TYPES = ['findings', 'findings_resources', 'findings_types']

ALL_ZONE_EXPORT_TYPES = ZONE_METRIC_EXPORT_TYPES + DATA_EXPORT_TYPES
ALL_ACCOUNT_EXPORT_TYPES = ACCOUNT_METRIC_EXPORT_TYPES + DATA_EXPORT_TYPES
